test/model/message_list_test.dart

```diff
@@ -162,5 +162,107 @@ void main() async {
+          userId: reaction.userId,
+          messageId: messageId,
+        );
+      }
+
+      test('add reaction', () async {
+        final originalMessage = eg.streamMessage(id: 243, stream: stream, reactions: []);
+        final messageList = await messageListViewWithMessages([originalMessage], stream, narrow);
+
+        final message = messageList.messages.single;
+        check(message).reactions.not(it()..jsonEquals([eg.unicodeEmojiReaction]));
+
+        bool listenersNotified = false;
+        messageList.addListener(() { listenersNotified = true; });
+
+        messageList.maybeUpdateMessageReactions(
+          mkEvent(eg.unicodeEmojiReaction, ReactionOp.add, originalMessage.id));
+
+        check(listenersNotified).isTrue();
+        check(messageList.messages.single)
+          ..identicalTo(message)
+          ..reactions.jsonEquals([eg.unicodeEmojiReaction]);
+      });
+
+      test('add reaction; message is not in list', () async {
+        final someMessage = eg.streamMessage(id: 1, reactions: []);
+        final messageList = await messageListViewWithMessages([someMessage], stream, narrow);
+        check(messageList.messages.single).reactions.jsonEquals([]);
+
+        bool listenersNotified = false;
+        messageList.addListener(() { listenersNotified = true; });
+
+        messageList.maybeUpdateMessageReactions(
+          mkEvent(eg.unicodeEmojiReaction, ReactionOp.add, 1000));
+
+        check(listenersNotified).isFalse();
+        check(messageList.messages.single).reactions.jsonEquals([]);
+      });
+
+      test('remove reaction', () async {
+        final eventReaction = Reaction(reactionType: ReactionType.unicodeEmoji,
+          emojiName: 'wave',                  emojiCode: '1f44b', userId: 1);
+
+        // Same emoji, different user. Not to be removed.
+        final reaction2 = Reaction.fromJson(eventReaction.toJson()..update('user_id', (_) => 2));
+
+        // Same user, different emoji. Not to be removed.
+        final reaction3 = Reaction.fromJson(eventReaction.toJson()
+          ..update('emoji_code', (_) => '1f6e0')
+          ..update('emoji_name', (_) => 'working_on_it'));
+
+        // Same user, same emojiCode, different emojiName. To be removed: servers
+        // key on user, message, reaction type, and emoji code, but not emoji name,
+        // so we follow that:
+        //   https://github.com/zulip/zulip-flutter/pull/256#discussion_r1284865099
+        final reaction4 = Reaction.fromJson(eventReaction.toJson()
+          ..update('emoji_name', (_) => 'tools'));
+
+        final originalMessage = eg.streamMessage(id: 243, stream: stream,
+          reactions: [eventReaction, reaction2, reaction3, reaction4]);
+        final messageList = await messageListViewWithMessages([originalMessage], stream, narrow);
+
+        final message = messageList.messages.single;
+        check(message).reactions.not(it()..jsonEquals([reaction2, reaction3]));
```

Comment:
Yeah, this test is definitely clearer by stating what it expects the starting point to be, rather than that it's different from the expected ending point. :-)

I think in fact this is clear enough if you just leave this check out — the definition of `originalMessage` is a few lines earlier and seems equally explicit.

Reply:
> I think in fact this is clear enough if you just leave this check out

Could the same be said for the other added tests, I wonder:

```dart
      test('add reaction', () async {
        final originalMessage = eg.streamMessage(stream: stream, reactions: []);
        final messageList = await messageListViewWithMessages([originalMessage], stream, narrow);

        final message = messageList.messages.single;
        check(message).reactions.jsonEquals([]);
```

```dart
      test('add reaction; message is not in list', () async {
        final someMessage = eg.streamMessage(id: 1, reactions: []);
        final messageList = await messageListViewWithMessages([someMessage], stream, narrow);
        check(messageList.messages.single).reactions.jsonEquals([]);
```

```dart
      test('remove reaction; message is not in list', () async {
        final someMessage = eg.streamMessage(id: 1, reactions: [eg.unicodeEmojiReaction]);
        final messageList = await messageListViewWithMessages([someMessage], stream, narrow);

        check(messageList.messages.single).reactions.jsonEquals([eg.unicodeEmojiReaction]);
```

Reply:
Yeah, I think so.